theano/sandbox/gpuarray/basic_ops.py

```diff
@@ -815,6 +815,44 @@ def perform(self, node, inp, out_):
                 raise ValueError("total size of new array must be unchanged")
         out[0] = x.reshape(tuple(shp))
 
+    def c_code_cache_version(self):
+        return (1,)
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        x, shape = inputs
+        output, = outputs
+        new_ndim = self.ndim
+        sdtype = node.inputs[1].type.dtype_specs()[1]
+        fail = sub['fail']
+        return """
+        assert (PyArray_NDIM(%(shape)s) == 1);
+        size_t new_dims[%(new_ndim)s];
+        int compute_axis = -1;
+        for (size_t i = 0; i < %(new_ndim)s; ++i)
+        {
+            new_dims[i] = ((%(sdtype)s*)(
+                    PyArray_BYTES(%(shape)s) +
+                    i * PyArray_STRIDES(%(shape)s)[0]))[0];
+            if (new_dims[i] == -1)
+            {
+                if (compute_axis != -1)
+                {
+                    PyErr_Format(PyExc_ValueError,
+                                 "GpuReshape: only one -1 is accepted "
+                                 "in the new shape, but got two at "
+                                 "indices %%d and %%zu.",
+                                 compute_axis, i);
+                    %(fail)s;
+                }
+                compute_axis = i;
+            }
+        }
+        
+        Py_XDECREF(%(output)s);
+        %(output)s = pygpu_reshape(%(x)s, %(new_ndim)s, new_dims,
+                                   GA_ANY_ORDER, 1, compute_axis);
```

Comment:
What happens when the new shape does not contain the same number of elements as the original shape? Is thia already verified by pygpu_reshape() and a reasonable error message is raised or should this be handled in this op?